### PR TITLE
hv_enforce_cpuid_msr_check: add new testcase

### PR DESCRIPTION
**Pull request by nickzhq**
hv_enforce_cpuid_msr_check: add new testcase
ID: 2051845
Signed-off-by: Houqi (Nick) Zuo hzuo@redhat.com

### CODE REVIEW

File: qemu/tests/cfg/hv_enforce_cpuid.cfg [hunk 1]
```diff
@@ -0,0 +1,21 @@
+- hv_enforce_cpuid:
+    only Linux
+    type = hv_enforce_cpuid
+    modprobe_cmd = "modprobe msr"
+    rdmsr_cmd = "rdmsr 0x40000002"
+    create_image = yes
+    remove_image = yes
```
Comment (zhencliu) [inline qemu/tests/cfg/hv_enforce_cpuid.cfg:7]: We can remove them, for we don't create new images

File: qemu/tests/hv_enforce_cpuid.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        # install msr-tools
+        status = session.cmd_status_output(params.get("msr_install_cmd"))
+        if status[0]:
+            test.error("Failed to install msr-tools")
+        # run msr-tools cmd
+        session.cmd_status_output(params.get("modprobe_cmd"), timeout=360)
+        res = session.cmd_status_output(params.get("rdmsr_cmd"), timeout=360)
+        # restore the subscription-manager
+        session.cmd_status_output(params.get("restore_submanager_cmd"))
+        vm.graceful_shutdown(timeout=360)
```
Comment (zhencliu) [inline qemu/tests/hv_enforce_cpuid.py:46]: vm.destroy() will also call graceful_shutdown, so we can remove it

File: qemu/tests/cfg/hv_enforce_cpuid.cfg [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    kill_vm_gracefully = yes
```
Comment (zhencliu) [inline qemu/tests/cfg/hv_enforce_cpuid.cfg:8]: Not a must IMO, vm.destroy() can make sure the previous vm can be destroyed.
this option seems to be used for preparation of starting a vm

File: qemu/tests/hv_enforce_cpuid.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+
```
Comment (zhencliu) [inline qemu/tests/hv_enforce_cpuid.py:1]: remove the empty line

File: qemu/tests/hv_enforce_cpuid.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        return: cmd output
+        """
+        vm = qemu_img_utils.boot_vm_with_images(test, params, env)
+        session = vm.wait_for_login(timeout=360)
+        # modify the subscription-manager
+        session.cmd_status_output(params.get("modify_submanager_cmd"))
+        # install repo
+        session.cmd_status_output(params.get("repo_install_cmd"))
+        # install msr-tools
+        status = session.cmd_status_output(params.get("msr_install_cmd"))
```
Comment (zhencliu) [inline qemu/tests/hv_enforce_cpuid.py:38]: if we don't want to check the output of a command, we can also use session.cmd_status(cmd), e.g. 
if session.cmd_status(cmd) != 0:
    # cmd failed

**Comment by menli820** [comment]
test failed with following errorz;
 (1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.cpu_hv_flags_msr_check.q35: ERROR: local variable 'session' referenced before assignment (521.24 s)
RESULTS    : PASS 0 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Comment by nickzhq** [comment]
> test failed with following error: (1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.cpu_hv_flags_msr_check.q35: ERROR: local variable 'session' referenced before assignment (521.24 s) RESULTS : PASS 0 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

This error was made by image. I'm looking at a new image.

**Comment by nickzhq** [comment]
@zhencliu @menli820  The update is coming! Please take a look. Thank you all!

File: qemu/tests/hv_enforce_cpuid_msr_check.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+        params session: the session of vm
+        return res: the result of output
+        """
+        session.cmd_output_safe(params.get("modprobe_cmd"), timeout=360)
+        res = session.cmd_output_safe(params.get("rdmsr_cmd"), timeout=360)
+        return res
+
+    error_context.context("The case starts...", test.log.info)
+    error_context.context("The case of cpu with hv flags: ", test.log.info)
```
Comment (menli820) [inline qemu/tests/hv_enforce_cpuid_msr_check.py:60]: better be:   
error_context.context("Boot the guest with 'hv-enforce-cpuid ", test.log.info)

File: qemu/tests/cfg/hv_enforce_cpuid_msr_check.cfg [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    rdmsr_cmd = "rdmsr 0x40000002"
+    not_preprocess = yes
+    required_qemu = [6.2.0,)
+    system_version_cmd = "cat /etc/redhat-release | grep -oE '[0-9]+\.[0-9]+'|cut -d '.' -f 1"
+    repo_install_cmd = "yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-${X_VER}.noarch.rpm"
+    msr_install_cmd = "yum install -y msr-tools"
+    modify_submanager_cmd = "sed -i -e 's/enabled=1/enabled=0/' /etc/yum/pluginconf.d/subscription-manager.conf"
+    restore_submanager_cmd = "sed -i -e 's/enabled=0/enabled=1/' /etc/yum/pluginconf.d/subscription-manager.conf"
+    cpu_model_flags_with_hv = 'hv-relaxed,hv-enforce-cpuid'
+    expect_result_with_hv = "rdmsr: CPU 0 cannot read MSR 0x40000002"
```
Comment (menli820) [inline qemu/tests/cfg/hv_enforce_cpuid_msr_check.cfg:14]: better be： expect_result_with_enforce

File: qemu/tests/cfg/hv_enforce_cpuid_msr_check.cfg [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    # we need add at least one hyperV flag to enable hyperV signature in guest to confirm following steps are valid.
+    cpu_model_flags_without_hv = 'hv-relaxed'
+    expect_result_without_hv = "0"
```
Comment (menli820) [inline qemu/tests/cfg/hv_enforce_cpuid_msr_check.cfg:17]: better be：
expect_result_without_enforce

**Comment by nickzhq** [comment]
@menli820 @zhencliu  Please check it out, thanks!

**Comment by menli820** [comment]
@nickzhq  the case error as following.
(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.hv_enforce_cpuid_msr_check.q35: ERROR: 'NoneType' object has no attribute 'cmd_output_safe' (5.26 s)

**Comment by nickzhq** [comment]
> RROR: 'NoneType' object has no attribute 'cmd_output_safe'

@menli820  Can you show me the command you used? Thanks!

**Comment by menli820** [comment]
> > RROR: 'NoneType' object has no attribute 'cmd_output_safe'
> 
> @menli820 Can you show me the command you used? Thanks!

python3 ConfigTest.py --testcase=hv_enforce_cpuid_msr_check --guestname=RHEL.9.0.0 --platform=x86_64 --machine=q35 --clone=no

**Comment by nickzhq** [comment]
(1/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
\ (1/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1492.47 s)
 (2/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.hv_enforce_cpuid_msr_check.q35: STARTED
/ (2/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.hv_enforce_cpuid_msr_check.q35: PASS (153.24 s)

@menli820  please check it out again

**Comment by menli820** [comment]
LGTM.

 (1/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1032.36 s)
 (2/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.hv_enforce_cpuid_msr_check.q35: PASS (209.50 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2022-03-06T20.35-782b90e/results.html

**Comment by nickzhq** [comment]
Here is an optimized code.  
Please take a look @zhencliu 
Thanks.

File: qemu/tests/cfg/hv_enforce_cpuid_msr_check.cfg [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    required_qemu = [6.2.0,)
+    RHEL.9:
+        repo_install_cmd = "yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm"
+    RHEL.8:
+        repo_install_cmd = "yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm"
+    RHEL.7:
+        repo_install_cmd = "yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm"
+    msr_install_cmd = "yum install -y msr-tools"
+    modify_submanager_cmd = "sed -i -e 's/enabled=1/enabled=0/' /etc/yum/pluginconf.d/subscription-manager.conf"
+    restore_submanager_cmd = "sed -i -e 's/enabled=0/enabled=1/' /etc/yum/pluginconf.d/subscription-manager.conf"
```
Comment (zhencliu) [inline qemu/tests/cfg/hv_enforce_cpuid_msr_check.cfg:16]: hi @nickzhq , I tried to install msr-tools, and found that we don't have to disable subscription-manager, msr-tool was installed successfully, would you pls make a double check for this? We can remove these if no need to do the change.

**Comment by nickzhq** [comment]
Hello @zhencliu  . The code has been changed! Please take a look!
Thanks a lot!

**Comment by zhencliu** [comment]
LGTM

@nickzhq pls run the test with --clone=yes to see if there is any issue

**Comment by nickzhq** [comment]
(1/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
/ (1/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1040.11 s)
 (2/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.hv_enforce_cpuid_msr_check.q35: STARTED
\ (2/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.hv_enforce_cpuid_msr_check.q35: PASS (165.79 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

Hello @zhencliu , here is result.

**Pull request merged**